Engine Hacks/SkillSystem/Internals/CalcLoops/WeaponUsabilityLoop/CanUnitWieldWeapon.s

```diff
@@ -0,0 +1,365 @@
+.thumb
+.align
+.equ LookupList,ItemTable+4
+.equ ExternalLoop,LookupList+4
+
+.macro blh to, reg
+    ldr \reg, =\to
+    mov lr, \reg
+    .short 0xF800
+.endm
+
+CanUnitWieldWeapon:
+push {r4-r7,r14}
+
+mov r4,r0 @r4= char struct pointer
+mov r5,r1 @r5= item halfword
+cmp r5,#0
+bne HasAWeapon
+b CannotWield
+
+HasAWeapon:
+mov r1,#0xFF
+and r1,r5 @get just item ID in r1
+mov r0,#0x24
+mul r0,r1
+ldr r1,ItemTable
+add r0,r1 @r0= item table data offset
+ldr r2,[r0,#8] @r2 = weapon ability word
+mov r0,#1
+and r0,r2 
+mov r3,r1 @r3 = item table offset
+cmp r0,#0 @check if the "is a weapon" bit is set
+bne IsAWeapon @if so, continue
+b CannotWield
+
+IsAWeapon:
+ldr r0,=#0x003D3C00 @word to be &ed to the item ability word 
+and r0,r2
+cmp r0,#0
+bne HasWeaponLocks
+b LockCheck7
+
+HasWeaponLocks:
+
+@this is terrible so let's redo it entirely
+@1. lookup list; each entry is 8 bytes, ties item ability bit to character/class ability bit in a word for each (this alone is 1/4 the size of the current function and is data that will likely be larger than the replacement for the current function!!)
```

Comment:
This doesn't seems to match what you're doing.

And about what you're doing (which is using byte/mask pairs):

I guess you did it that way to save bytes? (That seems unnecessary) Or maybe so that you can have a cleaner terminating value? Either way this is a bit eh imo.

What you could do I guess is just have each entry be a pair of bit indices in the item attr and char/class attr respectively. For example, the monster lock could just be defined like that:

```
    BYTE 0x14 0x1C // Monster Lock
```

And then your code could be simplified to the following:

```arm
LoopStart:
    ldrb r0, [r7]

    cmp r0, #0xFF
    beq LockCheck7 @ if we've reached the end, leave

    mov r1, #1 
    lsl r1, r0 @ get mask from bit index (1 << bit)

    tst r1, r3
    beq RestartLoop

    ldrb r0, [r7, #1]

    mov r1, #1 
    lsl r1, r0

    tst r1, r6
    beq CannotWield

RestartLoop:
    add r7, #2
    b LoopStart
```

idk. Maybe this would be a bit more obtuse to the typical user but I also really want to not encourage spreading the illusion that item attr and char/class attr are 4 different bytes (because they are not).

Reply:
What is described in the file was my initial idea, but then I wanted to save bytes because I thought it would go inline instead of a pointer to it in free space so I changed how it worked to take up less space; I changed it back to my initial idea since that 1. has ability words as words and 2. is even smaller space-wise